### PR TITLE
feat(cognito): user pool client - token validity

### LINKED ISSUE #11689: (cognito): add idTokenValidity to the UserPoolClient L2
> Originally from an internal ticket

On checking the CDK Typescript reference, I found that there are two constructs for defining a user pool app client. A high-level construct UserPoolClient [1] and a low-level construct CfnUserPoolClient [2]. The high-level constructor UserPoolClient does not have any provision to specify any token validity. However the low-level construct CfnUserPoolClient does have an option to provide the tokens validity, similar to CloudFormation template parameters.

### Use Case

Customer is currently using the higher-level constructs and he wants to know if there is there a plan to implement the token expires in the higher-level constructs as well. Because the lower-level constructs requires more (ugly) code.

### Other

[1] Class UserPoolClient:
https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-cognito.UserPoolClient.html

[2] Class CfnUserPoolClient:
https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-cognito.CfnUserPoolClient.html

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by jumi-dev**
Configuration to specify accessTokenValidity, idTokenValidity and refreshTokenValidity when a new user pool client is created in Cognito.

Closes #11689

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts [hunk 1]
```diff
@@ -248,6 +248,30 @@ export interface UserPoolClientOptions {
...
+   * Duration that specifies how long the ID token in the user pool client is valid.
+   * Values between 5 minutes and 1 day are valid. The duration can not be longer than the refresh token validity.
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cognito-userpoolclient.html#cfn-cognito-userpoolclient-idtokenvalidity
+   * @default Duration.minutes(60)
+   */
+  readonly idTokenValidity?: Duration;
+
+  /**
+   * Duration that specifies how long the refresh token in the user pool client is valid.
+   * Values between 60 minutes and 3650 days are valid.
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:262]:
```suggestion
   * Values between 60 minutes and 10 years are valid.
```

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts [hunk 2]
```diff
@@ -248,6 +248,30 @@ export interface UserPoolClientOptions {
    * registered with the user pool using the `UserPool.registerIdentityProvider()` API.
    */
   readonly supportedIdentityProviders?: UserPoolClientIdentityProvider[];
+
+  /**
+   * Duration that specifies how long the ID token in the user pool client is valid.
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:253]:
```suggestion
   * Validity of the ID token.
```

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts [hunk 3]
```diff
@@ -248,6 +248,30 @@ export interface UserPoolClientOptions {
...
+   * Values between 5 minutes and 1 day are valid. The duration can not be longer than the refresh token validity.
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cognito-userpoolclient.html#cfn-cognito-userpoolclient-idtokenvalidity
+   * @default Duration.minutes(60)
+   */
+  readonly idTokenValidity?: Duration;
+
+  /**
+   * Duration that specifies how long the refresh token in the user pool client is valid.
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:261]:
```suggestion
   * Validity of the refresh token.
```

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts [hunk 4]
```diff
@@ -248,6 +248,30 @@ export interface UserPoolClientOptions {
...
+   * Values between 60 minutes and 3650 days are valid.
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cognito-userpoolclient.html#cfn-cognito-userpoolclient-refreshtokenvalidity
+   * @default Duration.days(30)
+   */
+  readonly refreshTokenValidity?: Duration;
+
+  /**
+   * Duration that specifies how long the access token in the user pool client is valid.
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:269]:
```suggestion
   * Validity of the access token.
```

File: packages/@aws-cdk/aws-cognito/README.md [hunk 5]
```diff
@@ -550,6 +550,18 @@ pool.addClient('app-client', {
 });
 ```
 
+Each app client allows you to configure the expiration time for access token, ID token and refresh token. If no values are specified, the default values from Cognito are used.
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/README.md:553]:
```suggestion
In accordance with the OIDC open standard, Cognito user pool clients provide access tokens, ID tokens and refresh tokens.
More information is available at [Using Tokens with User Pools](https://docs.amazonaws.cn/en_us/cognito/latest/developerguide/amazon-cognito-user-pools-using-tokens-with-identity-providers.html).
The expiration time for these tokens can be configured as shown below.
```

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts [hunk 6]
```diff
@@ -248,6 +248,30 @@ export interface UserPoolClientOptions {
    * registered with the user pool using the `UserPool.registerIdentityProvider()` API.
    */
   readonly supportedIdentityProviders?: UserPoolClientIdentityProvider[];
+
+  /**
+   * Duration that specifies how long the ID token in the user pool client is valid.
+   * Values between 5 minutes and 1 day are valid. The duration can not be longer than the refresh token validity.
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cognito-userpoolclient.html#cfn-cognito-userpoolclient-idtokenvalidity
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:255]:
```suggestion
   * @see https://docs.amazonaws.cn/en_us/cognito/latest/developerguide/amazon-cognito-user-pools-using-tokens-with-identity-providers.html#amazon-cognito-user-pools-using-the-id-token
```

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts [hunk 7]
```diff
@@ -248,6 +248,30 @@ export interface UserPoolClientOptions {
...
+   * Values between 60 minutes and 3650 days are valid.
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cognito-userpoolclient.html#cfn-cognito-userpoolclient-refreshtokenvalidity
+   * @default Duration.days(30)
+   */
+  readonly refreshTokenValidity?: Duration;
+
+  /**
+   * Duration that specifies how long the access token in the user pool client is valid.
+   * Values between 5 minutes and 1 day are valid. The duration can not be longer than the refresh token validity.
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cognito-userpoolclient.html#cfn-cognito-userpoolclient-accesstokenvalidity
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:271]:
```suggestion
   * @see https://docs.amazonaws.cn/en_us/cognito/latest/developerguide/amazon-cognito-user-pools-using-tokens-with-identity-providers.html#amazon-cognito-user-pools-using-the-access-token
```

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts [hunk 8]
```diff
@@ -248,6 +248,30 @@ export interface UserPoolClientOptions {
...
+   * Values between 5 minutes and 1 day are valid. The duration can not be longer than the refresh token validity.
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cognito-userpoolclient.html#cfn-cognito-userpoolclient-idtokenvalidity
+   * @default Duration.minutes(60)
+   */
+  readonly idTokenValidity?: Duration;
+
+  /**
+   * Duration that specifies how long the refresh token in the user pool client is valid.
+   * Values between 60 minutes and 3650 days are valid.
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cognito-userpoolclient.html#cfn-cognito-userpoolclient-refreshtokenvalidity
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:263]:
```suggestion
   * @see https://docs.amazonaws.cn/en_us/cognito/latest/developerguide/amazon-cognito-user-pools-using-tokens-with-identity-providers.html#amazon-cognito-user-pools-using-the-refresh-token
```

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts [hunk 9]
```diff
@@ -322,6 +346,15 @@ export class UserPoolClient extends Resource implements IUserPoolClient {
...
 
+    let tokenValidityUnits : CfnUserPoolClient.TokenValidityUnitsProperty | undefined;
+    if (props.accessTokenValidity || props.idTokenValidity || props.refreshTokenValidity) {
+      tokenValidityUnits = {
+        idToken: props.idTokenValidity ? 'minutes' : undefined,
+        accessToken: props.accessTokenValidity ? 'minutes' : undefined,
+        refreshToken: props.refreshTokenValidity ? 'minutes' : undefined,
+      };
+    }
+
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:357]: Could you also add validations around the allowed limits, and corresponding unit tests?

Might be cleaner to move all of it and computing `tokenValidityUnits` into a private method.

File: packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts [hunk 10]
```diff
@@ -540,4 +540,64 @@ describe('User Pool Client', () => {
...
+      accessTokenValidity: Duration.minutes(60),
+    });
+    pool.addClient('Client3', {
+      userPoolClientName: 'Client3',
+      idTokenValidity: Duration.minutes(60),
+    });
+    pool.addClient('Client4', {
+      userPoolClientName: 'Client4',
+      refreshTokenValidity: Duration.days(30),
+    });
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts:567]: can you add another client with none of the token validity specified that asserts that the TokenValidityUnits is absent?

```ts
expect(stack).toHaveResource('AWS::Cognito::UserPoolClient', {
  ClientName: ...,
  TokenValidityUnits: ABSENT,
  IdTokenValidity: ABSENT,
  RefreshTokenValidity: ABSENT,
  AccessTokenValidity: ABSENT,
});
```

**Review by nija-at — DISMISSED** [review]
Thanks for submitting this PR 👍

See some of my comments below.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts [hunk 11]
```diff
@@ -416,4 +441,29 @@ export class UserPoolClient extends Resource implements IUserPoolClient {
...
+        refreshToken: props.refreshTokenValidity ? 'minutes' : undefined,
+      };
+    };
+
+    resource.idTokenValidity = props.idTokenValidity ? props.idTokenValidity.toMinutes() : undefined;
+    resource.refreshTokenValidity = props.refreshTokenValidity ? props.refreshTokenValidity.toMinutes() : undefined;
+    resource.accessTokenValidity = props.accessTokenValidity ? props.accessTokenValidity.toMinutes() : undefined;
+  }
+
+  private validateDurationInRangeOrUndefined(name: string, min: Duration, max: Duration, value?: Duration) {
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:463]: Keep it terse. Just `validateDuration` is sufficient.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts [hunk 12]
```diff
@@ -416,4 +441,29 @@ export class UserPoolClient extends Resource implements IUserPoolClient {
     if (providers.length === 0) { return undefined; }
     return Array.from(providers);
   }
+
+  private configureTokenValidity(resource: CfnUserPoolClient, props: UserPoolClientProps) {
+    this.validateDurationInRangeOrUndefined('idTokenValidity', Duration.minutes(5), Duration.days(1), props.idTokenValidity);
+    this.validateDurationInRangeOrUndefined('accessTokenValidity', Duration.minutes(5), Duration.days(1), props.accessTokenValidity);
+    this.validateDurationInRangeOrUndefined('refreshTokenValidity', Duration.minutes(60), Duration.days(10 * 365), props.refreshTokenValidity);
+
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:449]: Also validate that id tokens and access tokens are within the refresh token limits. You should be able to use the same `validateDuration...` method.

**Review by nija-at — CHANGES_REQUESTED** [review]
Thanks for making the changes. One last comment below. Looks good otherwise.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts (hunk 12 above)
Comment (jumi-dev) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:453]: Good remark, I added the additional validation and created the test cases for it.

**Comment by nija-at** [comment]
@jumi-dev - I've adjusted the tests slightly so that they are part of a separate `describe` block. Hope that's ok with you!

**Review by nija-at — APPROVED** [review]

**Pull request merged**